File: ch07/sec7.1.5/danc/max/maxelem.c
#include <stdio.h>

// Divide-and-Conquer: Find maximum using recursion
int findMax(int arr[], int low, int high) {
    // Base case: only one element
    if (low == high) {
        return arr[low];
    }

    // Divide: find middle
    int mid = low + (high - low) / 2;

    // Conquer: recursively find max in left and right halves
    int leftMax  = findMax(arr, low, mid);
    int rightMax = findMax(arr, mid + 1, high);

    // Combine: return the larger one
    return (leftMax > rightMax) ? leftMax : rightMax;
}

int main() {
    int arr[] = {13, 1, 45, 7, 92, 33, 58, 105, 22, 67};
    int n = sizeof(arr) / sizeof(arr[0]);

    int max = findMax(arr, 0, n - 1);

    printf("Array: ");
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\nMaximum element = %d\n", max);

    return 0;
}

File: ch07/sec7.1.5/danc/max/maxelem.py
import multiprocessing as mp

def find_max(arr, low, high):
    if low == high:
        return arr[low]
    
    mid = low + (high - low) // 2
    
    # Run left and right in parallel processes
    with mp.Pool(2) as pool:
        left_max = pool.apply_async(find_max, (arr, low, mid))
        right_max = pool.apply_async(find_max, (arr, mid + 1, high))
        results = [left_max.get(), right_max.get()]
    
    return max(results)

if __name__ == "__main__":
    arr = [13, 1, 45, 7, 92, 33, 58, 105, 22, 67]
    max_val = find_max(arr, 0, len(arr) - 1)
    print(f"Maximum element = {max_val}")
